### ISSUE TITLE
skopeo copy to containerd storage?

### ISSUE THREAD

**Issue by xwu2intel** [containers/image#1572]

**Comment by mtalexan** [comment]
I think this could be taken instead as a request for CRI-compatible socket communication as a format, much like how `docker-daemon` works.  

Currently skopeo supports cri-o by completely ignoring the CRI API and socket, and instead duplicating the container-storage code that cri-o also uses. This requires skopeo to read the same config files, come to the same conclusions, and have the same access permissions as cri-o.  This approach makes a lot of sense for podman, which is usually daemonless and therefore every instance of podman already has to do this on its own anyway.  Since cri-o intentionally uses the same underlying containers-storage code as podman, it makes sense for skopeo to intiailly support cri-o by not differentiating it from podman.  

However, cri-o _is_ a daemon tool, which opens up possibilities of how it might be configured (as a different user, with special lockdown/permissions, etc).  Skopeo isn't currently capable of dealing with most of these possibilities. Instead it currently requires that the `skopeo` command be run as the same user with the same permissions/context/etc as the cri-o daemon in order to access the same storage as cri-o, which is often prohibitively complicated/difficult/impossible.  
Adding support for cri-o via the CRI compatible socket eliminates these complexities and would incidentally also support containerd and other CRI-compatible tools.

**Comment by mtalexan** [comment]
> > I think this could be taken instead as a request for CRI-compatible socket communication as a format, much like how `docker-daemon` works.
> 
> I’m afraid this proposal has several incorrect assumptions. For example, here, the CRI API has no image import/export features at all, because K8s does not need anything of the kind.

I'm sure you're more of an expert on the API than me, but how do the `crictl` tools successfully import and export images to an arbitrary CRI-compatible daemon then?  


> > However, cri-o _is_ a daemon tool, which opens up possibilities of how it might be configured (as a different user, with special lockdown/permissions, etc). Skopeo isn't currently capable of dealing with most of these possibilities.
> 
> CRI-O has no concept of client permissions; there is only one CRI socket, and the user (typically Kubelet) can do anything — CRI-O _exists_ to serve Kubelet. So, any user the CRI CRI-O API has ~by definition all the same permissions as CRI-O, so it might just as well run in the same user/context.

I wasn't referring to the socket restrictions, I was talking about the cri-o daemon itself.  The cri-o daemon could be running under a dedicated system user for example, and/or be granted unique permissions.  In order for the independently executed container-storage module/code of skopeo to access the same storage as the cri-o daemon, skopeo would need to be executed with the same user and permissions as the cri-o daemon.  
Containers-storage is user-specific, that's the intended design that came from daemonless rootless podman originally.  But with cri-o being a deamon, you have an extremely high probability that the user it runs as isn't one you can easily use to run the `skopeo` command.

**Comment by mtrmac** [comment]
> but how do the `crictl` tools successfully import and export images to an arbitrary CRI-compatible daemon then?

I can’t see anything of that kind in https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md.

… are we talking about the same CRI and the same `crictl`?

**Comment by mtalexan** [comment]
> > but how do the `crictl` tools successfully import and export images to an arbitrary CRI-compatible daemon then?
> 
> I can’t see anything of that kind in https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md.
> 
> … are we talking about the same CRI and the same `crictl`?

Oh, you're right. I was thinking of `ctr`, which is part of the containerd tools. It seems they have a few non-CRI interfaces that they expose for doing image loading and saving.

**Comment by mtalexan** [comment]
TL;DR for future readers then:

The CRI interface doesn't support image loading or saving, so it can't be used for skopeo actions.  
CRI-O is only supported because its image storage mechanism is co-developed with skopeo and designed for independent access, while `containerd` is a completely unrelated tool.
